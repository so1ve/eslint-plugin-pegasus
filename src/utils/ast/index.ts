export * from "./call-or-new-expression";
export * from "./is-member-expression";
export * from "./is-method-call";
