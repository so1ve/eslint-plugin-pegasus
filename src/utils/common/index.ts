export * from "./is-node-matches";
export * from "./is-number";
export * from "./parentheses";
export * from "./rule";
