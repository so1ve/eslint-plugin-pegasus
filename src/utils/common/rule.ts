import type { TSESLint } from "@typescript-eslint/utils";

const isIterable = (object: any): object is Iterable<any> =>
	typeof object?.[Symbol.iterator] === "function";

class FixAbortError extends Error {}
const fixOptions = {
	abort() {
		throw new FixAbortError("Fix aborted.");
	},
};

export type AbortFunction = () => void;

type WrappedFixFunction = (
	fixer: TSESLint.RuleFixer,
	additional: { abort: AbortFunction },
) =>
	| IterableIterator<TSESLint.RuleFix>
	| TSESLint.RuleFix
	| readonly TSESLint.RuleFix[]
	| null;

export const wrapFixFunction =
	(fix: WrappedFixFunction): TSESLint.ReportFixFunction =>
	(fixer) => {
		const result = fix(fixer, fixOptions);

		if (isIterable(result)) {
			try {
				return [...result];
			} catch (error) {
				if (error instanceof FixAbortError) {
					return [];
				}

				throw error;
			}
		}

		return result;
	};
