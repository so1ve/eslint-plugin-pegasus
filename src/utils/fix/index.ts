export * from "./remove-member-expression-property";
export * from "./remove-method-call";
export * from "./replace-argument";
