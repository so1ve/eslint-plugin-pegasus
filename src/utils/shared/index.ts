export * from "./simple-array-search-rule";
