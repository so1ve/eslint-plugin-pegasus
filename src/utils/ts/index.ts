import typeutils from "@typescript-eslint/type-utils";
import type {
	ParserServicesWithTypeInformation,
	TSESTree,
} from "@typescript-eslint/typescript-estree";
import * as tsutils from "ts-api-utils";
import ts from "typescript";

export class TypeHelper {
	typeChecker: ts.TypeChecker;

	constructor(public services: ParserServicesWithTypeInformation) {
		this.typeChecker = services.program.getTypeChecker();
	}

	private getType(node: TSESTree.Node) {
		return typeutils.getConstrainedTypeAtLocation(this.services, node);
	}

	isTypeArrayTypeOrUnionOfArrayTypes(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeArrayTypeOrUnionOfArrayTypes(
			nodeType,
			this.typeChecker,
		);
	}

	isNumber(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return tsutils.isTypeFlagSet(nodeType, ts.TypeFlags.Number);
	}

	isAny(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeAnyType(nodeType);
	}
}
