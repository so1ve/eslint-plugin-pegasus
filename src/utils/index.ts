import { ESLintUtils } from "@typescript-eslint/utils";

export const createRule = ESLintUtils.RuleCreator(
	(ruleName) =>
		`https://github.com/so1ve/eslint-plugin-pegasus/blob/main/docs/rules/${ruleName}.md`,
);
