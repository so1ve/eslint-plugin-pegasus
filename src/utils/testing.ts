import { resolve } from "node:path";

import type { TSESLint } from "@typescript-eslint/utils";
import { expect } from "vitest";

import { RuleTester } from "../../vendor/rule-tester/src/RuleTester";

const fixturesDir = resolve(process.cwd(), "test", "fixtures");

export function getRuleTester() {
	const ruleTester = new RuleTester({
		parser: require.resolve("@typescript-eslint/parser"),
		parserOptions: {
			ecmaVersion: 2018,
			tsconfigRootDir: fixturesDir,
			project: "./tsconfig.json",
		},
	});

	return ruleTester;
}

export function test<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid,
		invalid,
	}: {
		valid: string[];
		invalid: TSESLint.InvalidTestCase<MessageIds, Options>[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid,
	});
}

export function testSnapshot<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid = [],
		invalid = [],
	}: {
		valid?: string[];
		invalid?: string[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid: invalid.map((i) => ({
			code: i,
			errors: null,
			onOutput: (output: string) => {
				expect(output).toMatchSnapshot();
			},
		})),
	});
}
