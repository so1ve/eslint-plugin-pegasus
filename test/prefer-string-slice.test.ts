import dedent from "dedent";

import rule, { RULE_NAME } from "../src/rules/prefer-string-slice";
import { test, testSnapshot } from "../src/utils/testing";

const errorsSubstr = [
	{
		messageId: "substr" as const,
	},
];
const errorsSubstring = [
	{
		messageId: "substring" as const,
	},
];

test(RULE_NAME, rule, {
	valid: [
		"const substr = foo.substr",
		"const substring = foo.substring",

		"foo.slice()",
		"foo.slice(0)",
		"foo.slice(1, 2)",
		"foo?.slice(1, 2)",
		"foo?.slice?.(1, 2)",
		"foo?.bar.baz.slice(1, 2)",
		"foo.slice(-3, -2)",
	],

	invalid: [
		{
			code: "foo.substr()",
			output: "foo.slice()",
			errors: errorsSubstr,
		},
		{
			code: "foo?.substr()",
			output: "foo?.slice()",
			errors: errorsSubstr,
		},
		{
			code: "foo.bar?.substring()",
			output: "foo.bar?.slice()",
			errors: errorsSubstring,
		},
		{
			code: "foo?.[0]?.substring()",
			output: "foo?.[0]?.slice()",
			errors: errorsSubstring,
		},
		{
			code: "foo.bar.substr?.()",
			output: "foo.bar.slice?.()",
			errors: errorsSubstr,
		},
		{
			code: "foo.bar?.substring?.()",
			output: "foo.bar?.slice?.()",
			errors: errorsSubstring,
		},
		{
			code: "foo.bar?.baz?.substr()",
			output: "foo.bar?.baz?.slice()",
			errors: errorsSubstr,
		},
		{
			code: "foo.bar?.baz.substring()",
			output: "foo.bar?.baz.slice()",
			errors: errorsSubstring,
		},
		{
			code: "foo.bar.baz?.substr()",
			output: "foo.bar.baz?.slice()",
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr()',
			output: '"foo".slice()',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1)',
			output: '"foo".slice(1)',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1, 2)',
			output: '"foo".slice(1, 3)',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(bar.length, Math.min(baz, 100))',
			output: '"foo".slice(bar.length, bar.length + Math.min(baz, 100))',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1, length)',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1, "abc".length)',
			output: '"foo".slice(1, 1 + "abc".length)',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr("1", 2)',
			errors: errorsSubstr,
		},
		{
			code: dedent`
				const length = 123;
				"foo".substr(1, length)
			`,
			output: dedent`
				const length = 123;
				"foo".slice(1, 1 + length)
			`,
			errors: errorsSubstr,
		},
		{
			code: dedent`
				const length = 123;
				"foo".substr(0, length)
			`,
			output: dedent`
				const length = 123;
				"foo".slice(0, Math.max(0, length))
			`,
			errors: errorsSubstr,
		},
		{
			code: dedent`
				const length = 123;
				"foo".substr('0', length)
			`,
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(0, -1)',
			output: '"foo".slice(0, 0)',
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(0, "foo".length)',
			output: '"foo".slice(0, "foo".length)',
			errors: errorsSubstr,
		},
		{
			code: dedent`
				const length = 123;
				"foo".substr(1, length - 4)
			`,
			output: dedent`
				const length = 123;
				"foo".slice(1, 1 + length - 4)
			`,
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1, length)',
			errors: errorsSubstr,
		},
		{
			code: dedent`
				const uri = 'foo';
				((uri || '')).substr(1)
			`,
			output: dedent`
				const uri = 'foo';
				((uri || '')).slice(1)
			`,
			errors: errorsSubstr,
		},

		{
			code: "foo.substr(start)",
			output: "foo.slice(start)",
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1)',
			output: '"foo".slice(1)',
			errors: errorsSubstr,
		},
		{
			code: "foo.substr(start, length)",
			errors: errorsSubstr,
		},
		{
			code: '"foo".substr(1, 2)',
			output: '"foo".slice(1, 3)',
			errors: errorsSubstr,
		},
		// Extra arguments
		{
			code: "foo.substr(1, 2, 3)",
			errors: errorsSubstr,
		},
		// #700
		{
			code: '"Sample".substr(0, "Sample".lastIndexOf("/"))',
			output: '"Sample".slice(0, "Sample".lastIndexOf("/"))',
			errors: errorsSubstr,
		},

		{
			code: "foo.substring()",
			output: "foo.slice()",
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring()',
			output: '"foo".slice()',
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(1)',
			output: '"foo".slice(1)',
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(1, 2)',
			output: '"foo".slice(1, 2)',
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(2, 1)',
			output: '"foo".slice(1, 2)',
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(-1, -5)',
			output: '"foo".slice(0, 0)',
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(-1, 2)',
			output: '"foo".slice(0, 2)',
			errors: errorsSubstring,
		},
		{
			code: 'const length = 1; "foo".substring(length)',
			output: 'const length = 1; "foo".slice(Math.max(0, length))', // TODO
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring("fo".length)',
			output: '"foo".slice("fo".length)',
			errors: errorsSubstring,
		},
		{
			code: 'const length = 1; "foo".substring(length, 0)',
			output: 'const length = 1; "foo".slice(0, Math.max(0, length))',
			errors: errorsSubstring,
		},

		{
			code: "foo.substring(start)",
			output: "foo.slice(Math.max(0, start))",
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(1)',
			output: '"foo".slice(1)',
			errors: errorsSubstring,
		},
		{
			code: "foo.substring(start, end)",
			errors: errorsSubstring,
		},
		{
			code: '"foo".substring(1, 3)',
			output: '"foo".slice(1, 3)',
			errors: errorsSubstring,
		},
		// Extra arguments
		{
			code: "foo.substring(1, 2, 3)",
			errors: errorsSubstring,
		},
	],
});

describe("prefer-string-slice", () => {
	testSnapshot(RULE_NAME, rule, {
		invalid: [
			dedent`
			/* 1 */ (( /* 2 */ 0 /* 3 */, /* 4 */ foo /* 5 */ )) /* 6 */
				. /* 7 */ substring /* 8 */ (
					/* 9 */ (( /* 10 */ bar /* 11 */ )) /* 12 */,
					/* 13 */ (( /* 14 */ 0 /* 15 */ )) /* 16 */,
					/* 17 */
				)
			/* 18 */
		`,
			"foo.substr(0, ...bar)",
			"foo.substr(...bar)",
			"foo.substr(0, (100, 1))",
			"foo.substr(0, 1, extraArgument)",
			"foo.substr((0, bar.length), (0, baz.length))",
			// TODO: Fix this
			// 'foo.substr(await 1, await 2)',
			"foo.substring((10, 1), 0)",
			"foo.substring(0, (10, 1))",
			"foo.substring(0, await 1)",
			"foo.substring((10, bar))",
		],
	});
});
