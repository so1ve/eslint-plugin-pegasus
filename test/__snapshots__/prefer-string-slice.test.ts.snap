// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-string-slice > prefer-string-slice > invalid > /* 1 */ (( /* 2 */ 0 /* 3 */, /* 4 */ foo /* 5 */ )) /* 6 */
\\u0009. /* 7 */ substring /* 8 */ (
\\u0009\\u0009/* 9 */ (( /* 10 */ bar /* 11 */ )) /* 12 */,
\\u0009\\u0009/* 13 */ (( /* 14 */ 0 /* 15 */ )) /* 16 */,
\\u0009\\u0009/* 17 */
\\u0009)
/* 18 */ 1`] = `
"/* 1 */ (( /* 2 */ 0 /* 3 */, /* 4 */ foo /* 5 */ )) /* 6 */
	. /* 7 */ slice /* 8 */ (
		/* 9 */ 0 /* 12 */,
		/* 13 */ Math.max(0, (( /* 10 */ bar /* 11 */ ))) /* 16 */,
		/* 17 */
	)
/* 18 */"
`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substr((0, bar.length), (0, baz.length)) 1`] = `"foo.slice((0, bar.length), (0, bar.length) + (0, baz.length))"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substr(...bar) 1`] = `"foo.substr(...bar)"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substr(0, (100, 1)) 1`] = `"foo.slice(0, Math.max(0, (100, 1)))"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substr(0, ...bar) 1`] = `"foo.substr(0, ...bar)"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substr(0, 1, extraArgument) 1`] = `"foo.substr(0, 1, extraArgument)"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substring((10, 1), 0) 1`] = `"foo.slice(0, Math.max(0, (10, 1)))"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substring((10, bar)) 1`] = `"foo.slice(Math.max(0, (10, bar)))"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substring(0, (10, 1)) 1`] = `"foo.slice(0, Math.max(0, (10, 1)))"`;

exports[`prefer-string-slice > prefer-string-slice > invalid > foo.substring(0, await 1) 1`] = `"foo.slice(0, Math.max(0, await 1))"`;
